oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.auth.oauth2;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.auth.oauth2.AccessTokenCapsule;
+import com.linecorp.armeria.common.auth.oauth2.InvalidClientException;
+import com.linecorp.armeria.common.auth.oauth2.RefreshAccessTokenRequest;
+import com.linecorp.armeria.common.auth.oauth2.TokenRequestException;
+import com.linecorp.armeria.common.auth.oauth2.UnsupportedMediaTypeException;
+import com.linecorp.armeria.common.util.Exceptions;
+
+/**
+ * Base implementation of OAuth 2.0 Access Token Grant flow to obtain Access Token.
+ * Implements Access Token loading, storing and refreshing.
+ */
+abstract class AbstractOAuth2AuthorizationGrant implements OAuth2AuthorizationGrant, AutoCloseable {
+
+    /**
+     * Holds a reference to the access token capsule.
+     */
+    private final AtomicReference<AccessTokenCapsule> tokenRef;
+
+    /**
+     * Executes obtain and refresh token operations serially on a separate thread.
+     */
+    private final ExecutorService serialExecutor;
+
+    private final RefreshAccessTokenRequest refreshRequest;
+
+    private final Duration refreshBefore;
+
+    @Nullable
+    private final Supplier<AccessTokenCapsule> tokenSupplier;
+    @Nullable
+    private final Consumer<AccessTokenCapsule> tokenConsumer;
+
+    AbstractOAuth2AuthorizationGrant(RefreshAccessTokenRequest refreshRequest, Duration refreshBefore,
+                                     @Nullable Supplier<AccessTokenCapsule> tokenSupplier,
+                                     @Nullable Consumer<AccessTokenCapsule> tokenConsumer) {
+        tokenRef = new AtomicReference<>();
+        serialExecutor = Executors.newSingleThreadExecutor();
```

Comment:
Do we have to create a new `ExecutorService` for every request? I think this might cause GC pressure in the high traffic. 🤔

Reply:
I think `expires` is not that short. So `RefreshAccessTokenRequest` might not be used for short live requests.

Reply:
There will be a dedicated instance created per `Grant`. Typically there should be only few `Grants` created ever - it's a resource that corresponds to a given `OAuth 2.0` authorization endpoint. Do you think there is still a risk?

Do you mean `refreshBefore`? `RefreshAccessTokenRequest` is connected to token expiration period and determined by the `OAuth 2` authorization endpoint. This validity period of the token is large enough and typically expressed in hours or days. There is not much sense to issue token just for a few seconds. Therefore, even default 1 minute to refresh before its expiry should be good enough for most of the cases. It can also be customized with the builder.
Also, the `Grant` lifespan should be independent from any client resource request.

Reply:
> There will be a dedicated instance created per Grant. Typically there should be only few Grants created ever - it's a resource that corresponds to a given OAuth 2.0 authorization endpoint. Do you think there is still a risk?

From the perspective of `OAuth2ClientCredentialsGrant`, that is true. Users do not need to create many `OAuth2ClientCredentialsGrant`s.

But `OAuth2ResourceOwnerPasswordCredentialsGrant` also inherits `AbstractOAuth2AuthorizationGrant` too. What if users want to handle various resource owners?
IIUC, based on the current implementation, they have to create `OAuth2ResourceOwnerPasswordCredentialsGrant` for every resource owners.

Reply:
Do you mean some multi-tenant use case, where the client serves multiple users? You are right, this will cause each `Grant` instance to have a dedicated single-thread executor. How do you suggest to make sharing of such thread/executor efficient and safe? I do not like an idea of injecting an external executor or a thread into the `Grant`. Should I simply make it static?

Reply:
Another alternative to `static` could be to wrap the `Executor` with some custom class (e.g. OAuth2AuthorizationGrantExecutor) and let it be injected via the `GrantBuilder`.